부산대 BE_원윤서 4주차(Step2)

yoonsu0325 opened the pull request:
step1에서 주신 피드백 반영했습니다!
옵션을 추가하면서 위시리스트도 Product가 아닌 Option을 담는 코드로 수정을 해야 할까요? 우선은 위시리스트 코드는 그대로 둔 상태입니다!

yoonsu0325 committed:
refactor : Category
getter setter 순서 변경

yoonsu0325 committed:
fix : CategoryController
getCategoriesByID -> getCategoryByID

yoonsu0325 committed:
feat : Option
옵션 클래스 생성
상품과 연관관계 매핑

yoonsu0325 committed:
feat : Product
옵션과 연관관계 매핑

yoonsu0325 committed:
feat: Option Repository

yoonsu0325 committed:
feat : Option DTOs
Option Response, Option Request DTO

yoonsu0325 committed:
feat : Option Controller
옵션 조회/추가/수정/삭제 API

yoonsu0325 committed:
feat : Option DTO
Option Request DTO

yoonsu0325 committed:
feat : Option Service
option의 비즈니스 로직 구현

yoonsu0325 committed:
feat : data.sql
기존 Product에 Option 추가

yoonsu0325 committed:
feat : Product Controller
Product 추가 시 기본으로 하나의 Option 생성

yoonsu0325 committed:
feat : Product Service
Product 추가 시 하나의 Option 생성

yoonsu0325 committed:
fix : GlobalExceptionHandler
위치 이동

yoonsu0325 committed:
fix : Option Requset, Option
Option Request의 name -> optionName

yoonsu0325 committed:
feat : Option Service, Option Repository
옵션명의 중복을 검사하는 로직 추가

yoonsu0325 committed:
fix : Option Controller
@Valid 추가

yoonsu0325 committed:
fix : Option Controller
@Valid 추가

yoonsu0325 committed:
feat : ProductController AddProduct.html
상품 추가 시 기본 옵션 추가

yoonsu0325 committed:
update README.md

yoonsu0325 committed:
fix : Option Service
옵션 수정 시에도 이름 중복 검사

deactivated-user commented on src/main/java/gift/product/ProductController.java:
```diff
@@ -25,13 +30,19 @@ public ProductController(ProductService productService, CategoryService category
     @PostMapping("/products/add")
-    public String addProduct(@Valid @ModelAttribute("newProduct") ProductDTO newProduct, BindingResult bindingResult, RedirectAttributes redirectAttributes) {
-        if (bindingResult.hasErrors()) {
+    public String addProduct(@Valid @ModelAttribute("newProduct") ProductDTO newProduct, BindingResult bindingResult1, @Valid @ModelAttribute("option") OptionRequest option, BindingResult bindingResult2, RedirectAttributes redirectAttributes) {
```
ProductDTO도 ProductResponse로 이름을 수정하는 건 어떨까요? (ProductRequest, ProductResponse)

deactivated-user commented on README.md:
```diff
@@ -1,16 +1,16 @@
-4. Category Repository
+1. Option class
+   - Product와 연관 관계 매핑
+   - wishList 수정
```
개인적으로는 위시리스트에도 옵션을 담도록 수정하는 것이 괜찮을 것 같습니다!

사용자가 제품에서 특정 옵션을 선택하여 위시리스트에 담아두면 나중에 위시리스트에서 바로 구매로 이어질 때, 옵션 선택 단계를 줄일 수 있습니다. 또한 위시리스트에서 바로 재고를 확인할 수도 있어 추가 기능 구현에 도움이 될 것 같습니다.

deactivated-user commented on src/main/java/gift/option/OptionController.java:
```diff
@@ -0,0 +1,36 @@
+        this.optionService = optionService;
+    }
+
+    @GetMapping("/{id}/options")
```
id보다는 productId로 명확하게 설명하는 것이 좋을 것 같습니다 (productId, optionId)

deactivated-user commented on src/main/java/gift/option/OptionService.java:
```diff
@@ -0,0 +1,53 @@
+
+    }
+
+    public boolean checkIfDuplicatedOption(Long id, OptionRequest option){
```
private 으로 둬도 좋을 것 같네요

deactivated-user commented on src/main/java/gift/option/OptionService.java:
```diff
@@ -0,0 +1,53 @@
+    }
+
+    public boolean checkIfDuplicatedOption(Long id, OptionRequest option){
+        return optionRepository.findByNameAndProductId(option.getOptionName(), id).isPresent();
```
exists는 조건에 맞는 데이터가 존재하는지만 확인하기 때문에 1건만 있어도 바로 반환하는데,
findBy는 전체 엔티티를 조회하기 때문에 exists로 바꾸는 것이 더 효율적일 것 같습니다!

deactivated-user commented on src/main/java/gift/option/OptionController.java:
```diff
@@ -0,0 +1,36 @@
+    @DeleteMapping("/{id}/options/{optionId}")
+    public void deleteProductOption(@PathVariable Long id, @PathVariable Long optionId){
+        optionService.deleteOption(id, optionId);
+    }
```
관리자 페이지와 사용자 페이지(실서비스)의 요구사항과 역할이 다르기 때문에.. 
관리자 페이지에서 쓰는 API와 사용자 페이지에서 쓰는 API가 분리되면 좋겠습니다.
섞여있는 경우 다른 개발자(서버, 클라 모두)가 해당 API의 역할을 이해하기 어렵습니다.

yoonsu0325 replied:
넵 우선 사용자 페이지는 상품ID로 옵션 조회와 옵션ID로 조회하는 API를 넣어두었고,
관리자 페이지에는 상품ID로 옵션 조회, 옵션 추가, 옵션 수정, 옵션 삭제 API로 나누는 방향으로 수정했습니다!

deactivated-user commented on src/main/java/gift/option/OptionRequest.java:
```diff
@@ -0,0 +1,41 @@
+    String optionName;
+    @NotNull(message = "수량은 공백으로 둘 수 없습니다.")
+    @Range(min = 1, max = 100000000, message = "최소 1개 이상 1억개 미만의 수량이 있어야 합니다.")
+    Long quantity;
```
여기는 기본 타입 long을 쓰는 것이 좋을 것 같습니다!

deactivated-user commented on src/main/java/gift/option/OptionService.java:
```diff
@@ -0,0 +1,53 @@
+        return optionRepository.findAllByProductId(id).stream().map(OptionResponse::new).toList();
+    }
+
+    public OptionResponse insertProductNewOption(Long id, OptionRequest optionRequest) {
```
일반적으로 하나의 제품에 여러 개의 옵션이 존재하는데요!

지금 옵션을 하나 추가할 때마다 API가 한 번 호출되는 구조라
여러 옵션을 한 번에 추가할 수 있으면 네트워크 요청 횟수를 줄여서 효율적일 것 같아요.

deactivated-user reviewed (changes requested):
step2 고생하셨습니다!
리뷰 코멘트 확인 부탁드려요~

deactivated-user commented on src/main/java/gift/option/OptionService.java:
```diff
@@ -0,0 +1,53 @@
+
+import java.util.List;
+
+@Service
```
옵션에 대한 테스트 코드도 있으면 좋겠습니다!

yoonsu0325 committed:
feat : Product Response, Request
Product DTO를 두개로 분리

yoonsu0325 committed:
fix : Option Service
checkIfDuplicatedOption -> private로 수정
checkIfDuplicatedOption -> findBy... 에서 existsBy...로 수정

yoonsu0325 committed:
fix : Option Repository
findBy -> existsBy

yoonsu0325 committed:
feat : Option
WishList와 연관관계 매핑

yoonsu0325 committed:
feat : Option과 연관관계 매핑

yoonsu0325 committed:
fix : Product
wishList와 연관관계 삭제

yoonsu0325 committed:
fix : WishList DTO, WishList Service
Product -> Option 연관관계 수정

yoonsu0325 committed:
fis : OptionRequest
Long quantity -> long quantity

yoonsu0325 committed:
feat : Option Controller, Option Manage Controller
컨트롤러를 관리자용 / 사용자용으로 분리

yoonsu0325 committed:
fix : Product Controller, Service
ProductDTO -> Product Requsest, Product Response

yoonsu0325 committed:
fix : Manager View
ProductDTO -> Product Requset, Product Response

yoonsu0325 committed:
feat : Product Request
생성자 수정

yoonsu0325 committed:
fix : JPA Test
테스트 수정

yoonsu0325 committed:
feat : Option Test
옵션 조회, 추가, 수정 테스트

deactivated-user commented on src/main/java/gift/wishList/WishList.java:
```diff
@@ -14,8 +15,8 @@ public class WishList {
     private User user;
     @ManyToOne(fetch = FetchType.LAZY)
-    @JoinColumn(name = "product_id")
-    private Product product;
```
앗 저는 WishList에서 Product와 Option을 둘 다 접근할 수 있어야 한다(*WishList는 User, Product, Option과 다대일 관계, 혹은 적어도 Product의 id 필드를 들고 있어야 한다)라고 생각해요.

서비스에서 유저가 위시리스트를 볼때 옵션뿐만 아니라 제품에 대한 정보도 보기를 원할 것 같아요.
<img width="1254" alt="image" src="https://github.com/user-attachments/assets/d0feac03-56c1-4787-8d8b-5535e265f07a">
장바구니를 생각했을 때 제품, 옵션이 모두 보이는 것 처럼요.

yoonsu0325 replied:
아하...! 제가 착각했네요 Product와 Option 둘 다 저장하는 방향으로 수정했습니다! 그리고 수정 하는김에 같은 위시리스트가 중복으로 저장되는 문제가 있었는데 그 부분도 수정했습니닷

deactivated-user commented on src/main/java/gift/product/ProductController.java:
```diff
@@ -25,21 +29,27 @@ public ProductController(ProductService productService, CategoryService category
     @PostMapping("/products/add")
-    public String addProduct(@Valid @ModelAttribute("newProduct") ProductDTO newProduct, BindingResult bindingResult, RedirectAttributes redirectAttributes) {
-        if (bindingResult.hasErrors()) {
+    public String addProduct(@Valid @ModelAttribute("newProduct") ProductRequest newProduct, BindingResult bindingResult1, @Valid @ModelAttribute("option") OptionRequest option, BindingResult bindingResult2, RedirectAttributes redirectAttributes) {
```
bindingResult1, 2가 각각 어떤 의미인지 잘 모르겠어요

yoonsu0325 replied:
지금 modelAttribute가 두개이고 각각 valid가 있어서 bindingResult를 따로 두었습니다!

deactivated-user commented on src/test/java/gift/OptionTest.java:
```diff
@@ -0,0 +1,99 @@
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SpringBootTest
+public class OptionTest {
```
👍

deactivated-user reviewed (changes requested):
리뷰 반영된 것 확인했습니다 :)

위시리스트 연관관계에서 의견 전달이 잘 안된 것 같아 추가 코멘트 달았습니다.
확인 부탁드려요!

yoonsu0325 committed:
feat : Product, WishList
연관관계 매핑

yoonsu0325 committed:
feat : WishList DTOs
WishList Response, WishList Request

yoonsu0325 committed:
feat : WishListRepository
findByUserAndOptionId

yoonsu0325 committed:
fix : WishListController
wishListRequest, wishListResponse 적용

yoonsu0325 committed:
feat : WishList Service
addWish
- Product 추가
- 이미 위시리스트가 존재하는 경우 count만 업데이트

deactivated-user reviewed (approved):
리뷰 코멘트 반영된 것 확인했습니다!
요청드린 사항이 많았는데 고생 많으셨어요.
머지하겠습니다 :)
